### PR TITLE
[Sampling] Capture masks from sampler support

### PR DESCRIPTION
**Pull request by ByronHsu**
## Problem

Sampling-mask replay must return the exact final support

```
S_t = {i | w_t[i] > 0}
```

and the sampled token logprob under that filtered distribution. The previous implementation reconstructed top-k/top-p/min-p support from the original probabilities after sampling, then repaired the reconstruction by adding the sampled token when necessary. Backend cutoff, tie, and filter-order semantics could make that reconstruction differ from the support actually used by the sampler. Sampled-token repair ensured only that the action was present; it could not restore other missing tokens or remove extra tokens, so replay could normalize over the wrong action space.

## Changes

- Add a private `_SamplingMaskCapture` carrying post-filter weights, an optional token-ID permutation, and the selected weight.
- Make `_sample_from_probs` return capture data only when sampling masks are requested.
- Reuse the exact sorted and filtered weights used by the PyTorch multinomial path.
- Reproduce FlashInfer effective joint top-k/top-p support with matching renormalization and intersection semantics; apply FlashInfer min-p capture sequentially as top-k → top-p → min-p.
- Represent greedy support as the selected-token singleton.
- Attach masks after token synchronization, define support strictly as `weights > 0`, and compute `log(w[y] / sum(w))` from the same capture.
- Treat a sampled token outside positive captured support as an internal invariant failure instead of repairing the mask.
- Remove post-hoc support reconstruction and sampled-token repair.
- Keep the existing synchronous ragged CPU return path and public response fields. Fixed-cap async D2H transport and overlap-scheduling support remain for a follow-up PR.

## Tests

- Registered sampling-mask suite in `test/registered/sampling/test_sampling_mask.py` covering FlashInfer and PyTorch.
- Exact support and selected-token logprob checks, including cutoff ties whose support exceeds `top_k`.
- `/generate`, chat completions, output alignment, rejection behavior, and seeded deterministic sampling coverage.
- H200 validation with `Qwen/Qwen2.5-1.5B-Instruct`: 9 tests passed.
- Python compilation, `git diff --check`, and relevant pre-commit hooks passed.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #33654612737](https://github.com/sgl-project/sglang/actions/runs/33654612737)
Latest PR Test (Extra): :white_check_mark: [Run #35022695411](https://github.com/sgl-project/sglang/actions/runs/35022695411)
Latest PR Test (AMD ROCm 10): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

File: python/sglang/srt/layers/sampler.py [hunk 1]
```diff
@@ -277,56 +308,81 @@ def _sample_from_probs(
...
                         filter_apply_order="joint",
                     )
+                    if return_sampling_mask:
+                        filtered_probs = probs
+                        if sampling_info.need_top_k_sampling:
+                            filtered_probs = top_k_renorm_prob(
+                                probs, sampling_info.top_ks
+                            )
+                        if sampling_info.need_top_p_sampling:
+                            top_p_probs = top_p_renorm_prob(probs, sampling_info.top_ps)
```
Comment (nanjiangwill) [inline python/sglang/srt/layers/sampler.py]: The overall approach makes sense: keep fused FlashInfer sampling unchanged and use the existing renormalization primitives only for opted-in capture. Could we document the required invariant here—that fused joint sampling and the separate top-k/top-p renormalization primitives share the same cutoff, tie, and joint-filter semantics? That invariant is the correctness boundary for this path and would be valuable to pin with a focused test.

File: test/registered/sampling/test_sampling_mask.py [hunk 2]
```diff
@@ -118,7 +118,7 @@ def test_generate_returns_sampling_mask(self):
             }
         )
         for sampling_mask in top_k_sampling_masks:
-            self.assertIn(len(sampling_mask), (_TOP_K, _TOP_K + 1))
+            self.assertGreaterEqual(len(sampling_mask), _TOP_K)
```
Comment (nanjiangwill) [inline test/registered/sampling/test_sampling_mask.py:301]: Keeping FlashInfer cutoff ties is the faithful behavior. It also means `top_k` is a sampling parameter, not a strict cardinality guarantee for the returned mask: threshold ties can produce more than `k` support entries, potentially the full vocabulary. Could we make sure the rest of this change preserves that responsibility boundary—the sampler returns exact support, while any fixed-capacity transport validates the realized length explicitly—and remove the now-unused `sampling_mask_max_top_k` sizing state?

File: python/sglang/srt/layers/sampler.py [hunk 3]
```diff
@@ -277,56 +308,81 @@ def _sample_from_probs(
...
+                            selected_weight=selected_weight,
+                        )
                 else:
                     batch_next_token_ids = top_k_top_p_sampling_from_probs(
                         probs.contiguous(),
                         sampling_info.top_ks,
                         sampling_info.top_ps,
                         filter_apply_order="joint",
                     )
+                    if return_sampling_mask:
```
Comment (nanjiangwill) [inline python/sglang/srt/layers/sampler.py:396]: Could the capture work be restricted to rows that requested a mask before running these full-vocabulary kernels? `return_sampling_mask` is batch-wide (`any(...)`), so one opted-in request currently causes FlashInfer to materialize filtered `[B, V]` weights for the entire batch, and `_attach_sampling_mask_to_output` later keeps only the requested rows. Ideally the opt-in cost should scale with the number of requested rows. When every row opts in, a direct path could also avoid the later full `index_select` copy.

File: python/sglang/srt/layers/sampler.py [hunk 4]
```diff
@@ -352,63 +408,94 @@ def _attach_sampling_mask_to_output(
...
+            selected_from_weights = torch.gather(
+                weights, 1, selected_positions
+            ).squeeze(1)
+            support_token_ids = token_ids
+
+        if selected_weight is None:
+            selected_weight = selected_from_weights
+
+        support = weights > 0
+        support_mass = torch.where(support, weights, torch.zeros_like(weights)).sum(
```
Comment (nanjiangwill) [inline python/sglang/srt/layers/sampler.py]: Since `weights` are already nonnegative and zero outside the captured support, would `weights.sum(dim=-1, dtype=torch.float32)` be equivalent here? That would keep the support predicate for packing while avoiding another dense float temporary on every opted-in decode step.

File: test/registered/sampling/test_sampling_mask.py [hunk 5]
```diff
@@ -175,17 +175,25 @@ def test_sampling_mask_matches_topk_logprobs(self):
...
+                cutoff = positional_support[-1][0]
+                reconstructed = {
+                    int(tid)
+                    for logprob, tid, _ in step_top_logprobs
+                    if math.exp(logprob) >= cutoff
+                }
+            else:
+                # PyTorch filters its sorted tensor by position.
+                reconstructed = {tid for _, tid in positional_support}
+            self.assertEqual(set(mask), reconstructed)
```
Comment (nanjiangwill) [inline test/registered/sampling/test_sampling_mask.py]: This live-model check is useful, but it cannot deterministically cover the tie boundary: the top 128 logprobs cannot reconstruct support larger than 128, and ordinary model outputs are unlikely to create an exact cutoff tie. Could we retain one focused derived-property test with synthetic tied probabilities, including the joint top-k/top-p boundary? That would directly verify that the fused sampling and renormalization primitives agree on the behavior this PR relies on.

Comment (ByronHsu) [reply test/registered/sampling/test_sampling_mask.py]: Added a focused synthetic joint-boundary test using probabilities `[0.4, 0.2, 0.2, 0.1, 0.1]`, `top_k=2`, and `top_p=0.45`. It verifies that tied cutoff entries produce support `{0, 1, 2}` and that fused joint samples remain within the support reconstructed by the renormalization primitives, without relying on live-model outputs or top-logprob truncation.

File: python/sglang/srt/layers/sampler.py (hunk 4 above)
Comment (ByronHsu) [reply python/sglang/srt/layers/sampler.py]: Applied in f22763062a: `support_mass` now uses `weights.sum(dim=-1, dtype=torch.float32)` directly, while the `weights > 0` predicate remains for sparse support packing. Captured weights are nonnegative with zeros outside support, so this preserves the result and avoids materializing the extra dense masked float tensor. The focused CUDA test and an attachment/logprob check both pass.

File: test/registered/sampling/test_sampling_mask.py (hunk 2 above)
Comment (ByronHsu) [reply test/registered/sampling/test_sampling_mask.py:301]: Addressed in a7c2ab6eee. This PR does not introduce a fixed-capacity transport; the normal output paths retain the full realized support in dynamic lists. The existing disaggregation metadata buffer is fixed-capacity and already validates `len(sampling_mask)` against its configured capacity before copying, raising explicitly on overflow rather than truncating. I also removed the now-unused `sampling_mask_max_top_k` field and its construction/merge bookkeeping.

File: python/sglang/srt/layers/sampler.py (hunk 1 above)
Comment (ByronHsu) [reply python/sglang/srt/layers/sampler.py]: Documented the correctness invariant in 51b3c59f9e at the fused joint sampling/capture boundary: the fused sampler and separate top-k/top-p renormalization primitives must share cutoff, tie, and joint-support semantics so the captured positive support matches the sampler action space. The focused synthetic regression with probabilities `[0.4, 0.2, 0.2, 0.1, 0.1]`, `top_k=2`, and `top_p=0.45` pins this tied joint boundary and passes.

File: python/sglang/srt/layers/sampler.py (hunk 3 above)
Comment (ByronHsu) [reply python/sglang/srt/layers/sampler.py:396]: Addressed in 1849430406. Sampling-mask captures now carry their original batch-row mapping and contain only opted-in rows. FlashInfer selects `[R, V]` probabilities and per-row parameters before the capture-only top-k/top-p renormalization kernels, while fused sampling remains unchanged on the full batch. `_attach_sampling_mask_to_output` consumes the compact capture directly, so the all-rows case also avoids the redundant full `index_select`. The PyTorch path still performs its sampling-required full-batch sort/filter, but compacts the returned capture before attachment. Focused CUDA coverage verifies the FlashInfer kernels receive `[R, V]`, partial outputs scatter to the correct batch rows, the all-row tie case remains faithful, and PyTorch captures are compact.
